### PR TITLE
Swallow errors while aborting remote transactions

### PR DESCRIPTION
**Pull request by velioglu**
DESCRIPTION: Fixes a bug that may cause crash while aborting transaction

Postgres doesn't allow erroring out within abort transaction callback, otherwise it crashes. To not to cause a crash, swallow any error for risky functions within transaction abort callback. 

Before that change, creating lots of partitions in a transaction block of MX cluster with coordinator node on the metadata was causing a crash. We were able to reproduce it on (some) local machine with following steps:

```
-- Add coordinator node to MX cluster with shouldhaveshards = true

CREATE TABLE ref(a int PRIMARY KEY);
SELECT create_reference_table('ref');

CREATE TABLE dist_1(a int PRIMARY KEY reference ref(a));
SELECT create_distributed_table('dist_1', 'a');

CREATE TABLE time_series_events (event_time timestamp, event int, user_id int REFERENCES dist_1(a)) partition by range (event_time);
SELECT create_distributed_table('time_series_events', 'user_id');

INSERT INTO ref SELECT i FROM generate_series(0,1000)i;
INSERT INTO dist_1 SELECT i FROM generate_series(0,1000)i;

BEGIN;
SELECT create_time_partitions(table_name:='time_series_events', partition_interval:= '1 day', end_at:=now() + '200 days', start_from:=now());
```

### CODE REVIEW

**Comment by SaitTalhaNisanci** [comment]
crush -> crash

File: src/backend/distributed/transaction/transaction_management.c [hunk 1]
```diff
@@ -325,7 +327,18 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 			/* handles both already prepared and open transactions */
 			if (CurrentCoordinatedTransactionState > COORD_TRANS_IDLE)
 			{
-				CoordinatedRemoteTransactionsAbort();
+				/*
+				 * Since CoordinateRemoteTransactionsAbort may cause an error and it is
+				 * not allowed to error out at that point, swallow the error if any.
```
Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_management.c:332]: It'd be good to describe the error we found.

In particular, CreateWaitEventSet may throw an error when out of file descriptors.

File: src/backend/distributed/transaction/transaction_management.c [hunk 2]
```diff
@@ -478,6 +491,74 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
...
+CloseAllInProgressConnections(void)
+{
+	dlist_iter iter;
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection,
+													  transactionNode,
+													  iter.cur);
+
+		CloseConnection(connection);
```
Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_management.c:511]: I think it'd be slightly nicer to set `connection->forceCloseAtTransactionEnd = true;` and let AfterXactConnectionHandling handle the closure.

File: src/backend/distributed/transaction/transaction_management.c [hunk 3]
```diff
@@ -478,6 +498,74 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
...
+	{
+		func();
+	}
+	PG_CATCH();
+	{
+		MemoryContextSwitchTo(savedContext);
+		ErrorData *edata = CopyErrorData();
+		FlushErrorState();
+
+		/* don't try to intercept PANIC or FATAL, let those breeze past us */
```
Comment (onurctirtir) [inline src/backend/distributed/transaction/transaction_management.c:551]: iirc, pg_try can't already catch panic / fatal

File: src/backend/distributed/transaction/transaction_management.c [hunk 4]
```diff
@@ -478,6 +494,69 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
...
+ForceAllInProgressConnectionsToClose(void)
+{
+	dlist_iter iter;
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection,
+													  transactionNode,
+													  iter.cur);
+
+		connection->forceCloseAtTransactionEnd = true;
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/transaction_management.c:511]: can we add a failure test such that we cover these lines?

Comment (onderkalaci) [reply src/backend/distributed/transaction/transaction_management.c:511]: Ok, this is not a simple place to add a test, so please ignore

**Review by marcocitus — APPROVED** [review]

**Pull request merged**